dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs

```diff
@@ -0,0 +1,473 @@
+                        AND {this.GetFullTableName(this._configuration.MemoryTableName)}.[collection] = @collection
+                ) AS [src]
+                ON [tgt].[memory_id] = [src].[id] AND [tgt].[vector_value_id] = [src].[vector_value_id]
+                WHEN MATCHED THEN
+                    UPDATE SET [tgt].[vector_value] = [src].[vector_value]
+                WHEN NOT MATCHED THEN
+                    INSERT ([memory_id], [vector_value_id], [vector_value])
+                    VALUES ([src].[id],
+                            [src].[vector_value_id],
+                            [src].[vector_value] )
+                ;";
+
+        cmd.Parameters.AddWithValue("@collection", collectionName);
+        cmd.Parameters.AddWithValue("@key", key);
+        cmd.Parameters.AddWithValue("@metadata", metadata ?? (object)DBNull.Value);
+        cmd.Parameters.AddWithValue("@embedding", embedding);
+        cmd.Parameters.AddWithValue("@timestamp", timestamp ?? (object)DBNull.Value);
+
+        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    private string GetFullTableName(string tableName)
+    {
+        return $"[{this._configuration.Schema}].[{tableName}]";
+    }
+
+    /// <inheritdoc />
+    private async Task<SqlServerMemoryEntry> ReadEntryAsync(SqlDataReader dataReader, bool withEmbedding, CancellationToken cancellationToken = default)
+    {
+        var entry = new SqlServerMemoryEntry();
+
+        entry.Id = dataReader.GetGuid(dataReader.GetOrdinal("id"));
+        entry.Key = dataReader.GetString(dataReader.GetOrdinal("key"));
+
+        if (!await dataReader.IsDBNullAsync(dataReader.GetOrdinal("metadata"), cancellationToken).ConfigureAwait(false))
+        {
+            entry.MetadataString = dataReader.GetString(dataReader.GetOrdinal("metadata"));
+        }
+
+        if (!await dataReader.IsDBNullAsync(dataReader.GetOrdinal("timestamp"), cancellationToken).ConfigureAwait(false))
+        {
+            entry.Timestamp = await dataReader.GetFieldValueAsync<DateTimeOffset?>(dataReader.GetOrdinal("timestamp"), cancellationToken).ConfigureAwait(false);
+        }
+
+        if (withEmbedding)
+        {
+            entry.Embedding = new ReadOnlyMemory<float>(JsonSerializer.Deserialize<IEnumerable<float>>(dataReader.GetString(dataReader.GetOrdinal("embedding")))!.ToArray());
+        }
+
+        return entry;
+    }
+
+    // Note: "_" is allowed in Postgres, but we normalize it to "-" for consistency with other DBs
+    private static readonly Regex s_replaceIndexNameCharsRegex = new(@"[\s|\\|/|.|_|:]");
+    private const string ValidSeparator = "-";
+
+    private static string NormalizeIndexName(string index)
+    {
+        index = s_replaceIndexNameCharsRegex.Replace(index.Trim().ToLowerInvariant(), ValidSeparator);
+
+        return index;
+    }
+
+    public void Dispose()
+    {
+        if (this._connection != null)
```

Comment:
I don't have a strong opinion about this, but calling `this._connection.Dispose();` is enough. Setting to `null` and doing the checks does not add any value (calling `Dispose` multiple times is fine).